crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -1129,3 +1129,80 @@ fn is_entry_primitive_type<E: fmt::Debug, S: StorageView<E>>(
     }
     Ok(true)
 }
+
+pub fn special_argument_validate(
+    bytes: &[u8],
+    idx: u16,
+    layout: SpecialArgumentLayout,
+) -> Result<(), ExecutionError> {
+    bcs::from_bytes_seed(&layout, bytes).map_err(|_| {
+        ExecutionError::new_with_source(
+            ExecutionErrorKind::command_argument_error(CommandArgumentError::TypeMismatch, idx),
+            format!("Function expects {layout} but provided argument's value does not match",),
+        )
+    })
+}
+
+impl<'d> serde::de::DeserializeSeed<'d> for &SpecialArgumentLayout {
+    type Value = ();
+    fn deserialize<D: serde::de::Deserializer<'d>>(
+        self,
+        deserializer: D,
+    ) -> Result<Self::Value, D::Error> {
+        use serde::de::Error;
+        match self {
+            SpecialArgumentLayout::Ascii => {
+                let s: &str = serde::Deserialize::deserialize(deserializer)?;
+                if !s.is_ascii() {
+                    Err(D::Error::custom("not an ascii string"))
+                } else {
+                    Ok(())
+                }
+            }
+            SpecialArgumentLayout::UTF8 => {
+                deserializer.deserialize_string(serde::de::IgnoredAny)?;
+                Ok(())
+            }
+            SpecialArgumentLayout::Option(layout) | SpecialArgumentLayout::Vector(layout) => {
+                deserializer.deserialize_seq(VectorElementVisitor(layout))
```

Comment:
@bmwill, I figure you might have some advice here :)

Reply:
You'll 100% want to have separate deserialize branches for option and vec in the unfortunate instance where the layout is Option and someone stuck more than 1 element in there. 

IIRC The bcs for option in move is equivalent to how a rust option is deserialized too so you could technically leverage the Option deserialize visitors if you wanted to.